### ISSUE TITLE
Rational Clock Constructors with variable intervalCounter argument

### ISSUE THREAD

**Issue by miscco** [modelica/ModelicaSpecification#2386]
Currently the specification reads 

> The first input argument, intervalCounter, is a clocked Component Expression (see Section 16.2.3) or a parameter expression of type Integer with min=0. The optional second argument resolution (default=1) is a parameter expression of type Integer with min=1 and unit “Hz”. If intervalCounter is a parameter expression with value zero, the period of the clock is derived by clock inference, see Section 16.7.5.

The unclear part is the last sentence
> If intervalCounter is a **parameter** expression with value zero, the period of the clock is derived by clock inference, see Section 16.7.5.

What happens if we have a clocked component expression that evaluates to zero, e.g
```
model ClockConstructorRationalIntervalInferenceInvalid
	Integer intervalCounter = 0;
	Clock rationalClock = Clock(intervalCounter); // Only parameter expressions are allowed for clock inference
end ClockConstructorRationalIntervalInferenceInvalid;
```

I dont see any compelling reason why it should be invalid. More importantly one could build systems with variable interval where the frequency is driven by one component, e.g.
```
model MultipleClocksVariableInterval
	Integer frequency1(start = 2);
	Integer frequency2(start = 0);
	Clock baseClock1 = Clock(frequency1, 10);
	Clock baseClock2 = Clock(frequency2, 10);
        Boolean someCondition;
	equation
                when sample(someCondition, baseClock1) then
                     frequency1 = 0;
                end when;
                when sample(someCondition, baseClock2) then
                     frequency2 = 2;
                end when;
end MultipleClocksVariableInterval;
```

This looks like it would be a valid use case. Consequently I would propose that we rephrase the last sentence to:

> If intervalCounter evaluates to zero, the period of the clock is derived by clock inference, see Section 16.7.5.

Note that this would also imply that we have to perform clock inference during run time. However this is not a realy problem, as every factor from the subclock conversion operators is parametric and needs to be determined only once. Consequently the "runtime clock inference" would be quite cheap as we would only have to check for consistence.

**Comment by HansOlsson** [comment]
I agree that it would be good to have it clearer, but I'm not sure that this will work well.

> Note that this would also imply that we have to perform clock inference during run time. However this is not a realy problem, as every factor from the subclock conversion operators is parametric and needs to be determined only once. Consequently the "runtime clock inference" would be quite cheap as we would only have to check for consistence.

As I see it there are two issues:
* If we can derive the subclock conversion from other parts doesn't that imply that this subclock conversion is highly constrained making this is an unusal case? Can we build any realistic models using this?
* For one operator I agree that it doesn't seem bad, but what happens if we have n subclock conversions operators using this and any combination of them could have non-parametric zero? To me it seems as if there 2^n^ combinations that need to be considered.

**Comment by miscco** [comment]
@Hans I think you canfuse the clock constructor with the subclock conversion operators. The arguments of the subclock conversion operators have parametric variability. Only the intervalCounter argument of the rational/real interval clock constructor have variable variability  

I think, that there are actually two different phases of clock inference

1. Inference of factors of sublock conversion operators. Everything that is declared as 0 we need to determine.

2. Determine the base interval of the base partition. Here we need to also check whether different rational interval clocks are consistent.

The first step is only needed at the beginning of the simulation, as the inputs to the subclock conversion operators have parametric variability and cannot ever change during the simulation. Consequently, the only thing we would have to do is determining the base interval of the base partition which is quite cheap

**Comment by HansOlsson** [comment]
My impression is that it would allow something like:

```
model Unnamed
    Integer nextInterval(start=2);    
    Real y2(start=0); 
equation
  when Clock(nextInterval, 1000) then   
         nextInterval = mod(previous(nextInterval) + 1,10);     
         y2 = previous(y2) + 1;   
  end when;
end Unnamed;
```
Where interval is periodically 0.

**Comment by miscco** [comment]
Technically this can already happen at which point the implementation can decide what base interval it uses

**Comment by miscco** [comment]
To be more precise we already have the same situation with
```
when Clock() then
    ...
end when;
```

**Comment by HansOlsson** [comment]
> Technically this can already happen at which point the implementation can decide what base interval it uses

Or we say that non-parameter interval with value zero is an error.

Note that in case of a non-parameter interval we must compute the new value of interval at each clock tick, I assume that means that interval and the generated clock must be on the same clock - which really restricts the possibilities, and I don't see how we can build something based on this.

Obviously we could allow non-parameter intervals that don't change, such as "Integer interval=0;" but I don't see a major benefit.

**Comment by miscco** [comment]
In principle I would guess, that we either want to completely dissalow variable Intervals to be of Value 0 or allow them always. Everything in between leads to problems.

> Note that in case of a non-parameter interval we must compute the new value of interval at each clock tick

This is already the case for every non-periodic clock. The moment interval is not a parameter expression we need to do this anyway. 

Even worse, the moment you have more than one nonperiodic rational interval clock you have to check whether each pair of rational interval base clocks is still consistent. One can obviously get that down to N-1 comparisons for N clocks but still this is technically quite hard, as you accidentally might end up trying to check 2 clocks with interval = 0.

Consequently you must perform a clock inference (the second part of it) whenever the interval of a base clock changes.

**Comment by HansOlsson** [comment]
> In principle I would guess, that we either want to completely dissalow variable Intervals to be of Value 0 or allow them always. Everything in between leads to problems.

TL;DR: Always is disallowing 0 is probably the best solution.

Full explanation:
If we compare with `subSample` we also have a way of inferring the sub-sampling factor, similarly as interval=0 infers the rational clock. However, it is a separate call `subSample(u)` instead of `subSample(u, factor=0)`. 

We can do something similar for clocks - we can just use `Clock()` instead of `Clock(intervalCounter=0)`, so we don't really need the second variant it seems.

However, those two calls are not necessarily the same - the specification needs to be clarified. Clearly `Clock()` can infer any kind of clock whereas it is described as if `Clock(intervalCounter=0)` infers the period, which seems like an odd statement if we infer that it is an event clock - or a non-periodic clock. It's also unclear if the resolution matters in that case.

Assuming `Clock(intervalCounter=0)` can infer everything then it is just the same as `Clock()`, and we could deprecate it without any problems. I would assume it is mostly used in components (but not in Modelica_Synchronous) - where we have a `intervalCounter` as a parameter, documented as "set to zero to infer" - and that could be handled by some if-statements in that model.

**Comment by miscco** [comment]
I would disagree that `Clock(intervalCounter=0)` is the same as `Clock()

`Clock()` says "give me the clock i belong to"
`Clock(intervalCounter=0)` says " Create me a new clock"

Both are fundamentally different. 

I think that ``Clock(intervalCounter=0)`` is only a problem because we can have multiple rational interval clocks in the same base partition. Therefore we need to check whether these base clocks are consistent. The "intervalCounter=0" case doesnt do anything here. 

The moment we have multiple rational interval base clocks with variable `intervalCounter` argument we must check for consistence. I dont know about your implementation but I found the easiest way to do that generically is via a reduced clock inference that does not infere factors but simply checks whether the corresponding graph is consistent.

**Comment by miscco** [comment]
Consider the following example:
```
model MultipleBaseClocks
	//
	//             Clock(1, frequency1)
	//                      |	  
	//                      | subSample(2)	
	//                      ↓	
	//             Clock(1, frequency2)   
	//
       Integer frequency1(start = 1);
       Integer frequency2(start = 2);
       Real mixedVar;
	equation
		mixedVar = sample(time, Clock(1, frequency1)) + subSample(sample(time, Clock(1, frequency2), 2));
end MultipleBaseClocks;
```

Here we end up with a single base partition that has 2 rational interval clocks. Both rational interval clocks have variable frequency. Whenever one of `{frequency1, frequency2}` changes we need to ensure that the other is consistent with the constraint that `2 * frequency1 == frequency2`

This has nothing to do with one of them evaluating to 0. This is just multiple rational interval clocks in the same base partititon. The only difference the `intervalCounter=0` case would make is that you assign the correct value to the infered rational interval clock rather than checking whether it is correct.

**Comment by HansOlsson** [comment]
> I would disagree that Clock(intervalCounter=0) is the same as Clock()
> Clock() says "give me the clock i belong to"
> Clock(intervalCounter=0) says " Create me a new clock"
> Both are fundamentally different.

That's one way of looking at them.

Another is to say that "Clock()" gives a new Clock that will be inferred (technically this is what the specification states), and "Clock(intervalCounter=0)" gives a new rational clock that will be inferred.

In that case they are more similar, and it isn't clear whether the second variant is needed - and if so if that it is best way to describe it.

**Comment by HansOlsson** [comment]
> Both rational interval clocks have variable frequency. Whenever one of {frequency1, frequency2} changes we need to ensure that the other is consistent with the constraint that 2 * frequency1 == frequency2

I understand (even though technically the resolution must be a parameter-expression.)

But I would argue that we should split the list in 16.3 Clock constructors as follows:

1a. One or more Rational interval clocks with parametric intervalCounter, provided they are consistent with each other
1b. Exactly one Rational interval clock where the intervalCounter is a clocked expression.

Having to check that two clocked expressions are always consistent is something we normally avoid in the specification - and I don't see a major need for it. (Disclaimer: As far as I understand we already have this restriction in the Dymola kernel and people don't complain.)

**Comment by miscco** [comment]
I am not really sure whether we should really dissallow it. 

I could imagine, that there are some models that would like to increase intervalCounter so that for example the choosen solverMethod is still numerically stable. 

This is a real corner case that will not be used often. We just need to ensure that the "normal" care is still efficient

**Comment by HansOlsson** [comment]
Need to analyze more.

**Comment by HansOlsson** [comment]
Proposal:

I don't see that it makes sense to dynamically switch between inferring the clock and giving it, so I would say add:

If intervalCounter is a not a parameter expression its value must be greater than zero.

**Comment by HansOlsson** [comment]
Agreement, but should improve formulation.
